Add IOBufOutputStream

oerling opened the pull request:
Adds an OutputStream that allows retrieving the content as
folly::IOBufs. This is for use in partitioned output to get zero copy
message passing between Velox and the exchange server in presto_cpp.

Changes the Position type of ByteStream to a number so as to be
compatible with C++ streams.

majetideepak commented on velox/common/memory/ByteStream.cpp:
```diff
@@ -17,9 +17,44 @@
+  if (current_ && current_->position > current_->fill) {
+    current_->fill = current_->position;
+  }
+  if (ranges_.empty() && position == 0) {
```
should this be `(ranges_.empty() || position == 0)`?

majetideepak commented on velox/common/memory/ByteStream.cpp:
```diff
@@ -31,4 +66,55 @@ void ByteStream::flush(OutputStream* out) {
+  if (current_ && current_ != &ranges_.back()) {
+    ++current_;
+    current_->fill = current_->position;
+    if (current_->fill < current_->position) {
```
This block of code is not clear. We have `current_->fill = current_->position;` on line 75 and then 
`if (current_->fill < current_->position)` on line 76 without any body.

mbasmanova replied:
Like Deepak pointed out this "if" statement is not doing anything. Please, remove.

majetideepak reviewed (commented):
@oerling I made a high-level pass. Left a couple of comments. Thanks.

mbasmanova commented:
Assuming this PR is no longer needed. Feel free to re-open if needed.

mbasmanova commented on velox/common/memory/ByteStream.cpp:
```diff
@@ -17,9 +17,44 @@
+void ByteStream::seekp(std::streampos position) {
+  int64_t toSkip = position;
+  if (current_ && current_->position > current_->fill) {
+    current_->fill = current_->position;
```
I find this logic confusing. Shouldn't "fill" be updated when "position" is updated?

mbasmanova commented on velox/common/memory/ByteStream.cpp:
```diff
@@ -31,4 +66,55 @@ void ByteStream::flush(OutputStream* out) {
+  current_ = &ranges_.back();
+  arena_->newRange(bytes, current_);
+}
+namespace {
```
nit: add an empty line before this

mbasmanova commented on velox/common/memory/ByteStream.cpp:
```diff
@@ -31,4 +66,55 @@ void ByteStream::flush(OutputStream* out) {
+} // namespace
+
+std::unique_ptr<folly::IOBuf> IOBufOutputStream::getIOBuf() {
+  // Make an IOBuf for each range. The The IOBufs keep shared ownership of
```
typo: "The The" -> "The"

mbasmanova reviewed (changes requested):
@oerling Some comments below.

oerling commented:
In seekp, it should be ranges_.empty() && position == 0. If we do a seekp into a stream without content, the only legitimate offset is 0. Anything else should error out.

 

Fixed extend. Extend is covered in the test, where we write for the first time and then rewrite everything.

 

Added some comments about the relation of members of ByteRange.

 

 

From: Deepak Majeti ***@***.***> 
Sent: Monday, March 7, 2022 6:37 AM
To: facebookincubator/velox ***@***.***>
Cc: oerling ***@***.***>; Mention ***@***.***>
Subject: Re: [facebookincubator/velox] Add IOBufOutputStream (PR #1123)

 

@majetideepak commented on this pull request.

@oerling <https://github.com/oerling>  I made a high-level pass. Left a couple of comments. Thanks.

  _____  

In velox/common/memory/ByteStream.cpp <https://github.com/facebookincubator/velox/pull/1123#discussion_r820752529> :

> +  int64_t size = 0;
+  for (auto& range : ranges_) {
+    if (&range == current_) {
+      return current_->position + size;
+    }
+    size += range.numValues();
+  }
+  VELOX_FAIL("ByteStream 'current_' is not in 'ranges_'.");
+}
+
+void ByteStream::seekp(std::streampos position) {
+  int64_t toSkip = position;
+  if (current_ && current_->position > current_->fill) {
+    current_->fill = current_->position;
+  }
+  if (ranges_.empty() && position == 0) {

should this be (ranges_.empty() || position == 0)?

  _____  

In velox/common/memory/ByteStream.cpp <https://github.com/facebookincubator/velox/pull/1123#discussion_r820758340> :

> @@ -31,4 +66,55 @@ void ByteStream::flush(OutputStream* out) {
   }
 }
 
+void ByteStream::extend(int32_t bytes) {
+  if (current_) {
+    current_->fill = current_->position;
+  }
+  if (current_ && current_ != &ranges_.back()) {
+    ++current_;
+    current_->fill = current_->position;
+    if (current_->fill < current_->position) {

This block of code is not clear. We have current_->fill = current_->position; on line 75 and then
if (current_->fill < current_->position) on line 76 without any body.

  _____  

In velox/common/memory/ByteStream.cpp <https://github.com/facebookincubator/velox/pull/1123#discussion_r820763667> :

> @@ -31,4 +66,55 @@ void ByteStream::flush(OutputStream* out) {
   }
 }
 
+void ByteStream::extend(int32_t bytes) {
+  if (current_) {
+    current_->fill = current_->position;
+  }
+  if (current_ && current_ != &ranges_.back()) {
+    ++current_;
+    current_->fill = current_->position;
+    if (current_->fill < current_->position) {

Probably add a test for ByteStream::extend

—
Reply to this email directly, view it on GitHub <https://github.com/facebookincubator/velox/pull/1123#pullrequestreview-901757980> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKPPPT6MS64AICSCIINKAP3U6YIATANCNFSM5P36OSZA> .
Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>  or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub> . 
You are receiving this because you were mentioned.Message ID: ***@***.***>

oerling commented:
Thank you, Masha.  Edited as requested.

 

For ByteRange, we can look at it like a folly::Range, spo buffer and size (count of addressable elements) are the right names. Fill only makes sense for the last one. So I put this in ByteStream. The point is that all but the last range are filled to size and the last range is filled to lastRangeEnd_. A seekp can set the writing position to something else, so we must keep track of the end (e.g. file length) separately. The use case is when writing stuff into Page serialization and then seeking back to the start to fill in size and checksum.

 

 

The end of last range must be set on seekp and is convenient to set before flush.

oerling commented:
Updated the follow up item 1127. would you take a look?

mbasmanova commented on velox/common/memory/ByteStream.h:
```diff
@@ -18,17 +18,19 @@
 namespace facebook::velox {
 
 struct ByteRange {
  public:
```
nit: public: is not needed as member variables of a struct are public by default

mbasmanova commented on velox/common/memory/ByteStream.h:
```diff
@@ -121,36 +134,40 @@ class ByteStream {
+  void seekp(std::streampos position);
 
+  // Returns the size written into ranges_. This is the sum of the
+  // capcities of non-last ranges + the write position of the last
```
typo: capcities -> capacities

mbasmanova commented on velox/common/memory/ByteStream.h:
```diff
@@ -40,47 +42,58 @@ class OutputStreamListener {
+// in hash tables. The stream is seekable and supports overwriting of
+// previous content, for example, writing a message body and then
+// seeking to start to write a length header.
 class ByteStream {
```
It is super confusing that this class is used as both input and output stream. It would be nice to split it up or document clearly which methods are supposed to be used for reading and which are for writing.

mbasmanova reviewed (approved):
@oerling Thank you, Orri.
